### PR TITLE
fix(fetch): integration with MockAgent

### LINKED ISSUE #1121: Incompatibility between undici fetch and MockAgent
## Bug Description

When using fetch and attempting to intercept requests via the MockAgent, the object passed to the body request matcher is an async generator rather than a string as documented [here](https://undici.nodejs.org/#/docs/api/MockPool?id=parameter-mockpoolinterceptoptions). When using request, the behavior is as documented.

### Code
```
import { request, fetch, MockAgent, setGlobalDispatcher } from "undici";

export const mockAgent = new MockAgent();
setGlobalDispatcher(mockAgent);

const mockPool = mockAgent.get("http://localhost:3000");

mockPool.intercept({
    path: "/test", method: "POST", body: (value) => {
        console.log(value);
        return value === "request body";
    }
}).reply(200, { success: true }).persist();

(async () => {
    try {
    await fetch("http://localhost:3000/test", { method: "POST", body: "request body" });
    console.error('fetch succeeded');
    } catch(err) {
        console.error('fetch failed');
        console.error(err);
    }
})().then(async () => {
    try {
        await request("http://localhost:3000/test", { method: "POST", body: "request body" });
        console.error('request succeeded');
        } catch(err) {
            console.error('request failed');
            console.error(err);
        }
});

```

## Expected Behavior

MockAgent should behave identically in both request and fetch.

### Console output from above code
```
Object [AsyncGenerator] {}
(node:47299) ExperimentalWarning: stream/web is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
fetch failed
TypeError: fetch failed
    at Object.processResponse (/Users/me/undici-bug/node_modules/undici/lib/fetch/index.js:176:23)
    at fetchFinale (/Users/me/undici-bug/node_modules/undici/lib/fetch/index.js:720:17)
    at EventEmitter.mainFetch (/Users/me/undici-bug/node_modules/undici/lib/fetch/index.js:695:5)
    at processTicksAndRejections (node:internal/process/task_queues:96:5) {
  cause: Error: connect ECONNREFUSED 127.0.0.1:3000
      at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1161:16) {
    errno: -61,
    code: 'ECONNREFUSED',
    syscall: 'connect',
    address: '127.0.0.1',
    port: 3000
  }
}
request body
request succeeded
```
## Environment

mac OS 10.15.7
Node 16.13.1

### PR DESCRIPTION
**Pull request by gdorsi**
fixes #1121

### CODE REVIEW

File: lib/fetch/index.js [hunk 1]
```diff
@@ -1621,7 +1622,7 @@ function httpNetworkFetch (
         path: url.pathname + url.search,
         origin: url.origin,
         method: request.method,
-        body,
+        body: context.dispatcher[kIsMockActive] ? request.body && request.body.source : body,
```
Comment (gdorsi) [inline lib/fetch/index.js:1625]: I tried a cleaner solution [here](https://github.com/nodejs/undici/pull/1149/commits/a04da434d3c67cb490c6276a428b0049184a6e40) but would've made things complicated

**Review by mcollina — APPROVED** [review]
lgtm

**Review by ronag — APPROVED** [review]

**Comment by mcollina** [comment]
It's failing on Node v12 on all platforms, can you take a look?

File: lib/mock/mock-utils.js [hunk 2]
```diff
@@ -33,7 +33,7 @@ function matchHeaders (mockDispatch, headers) {
   }
 
   for (const [matchHeaderName, matchHeaderValue] of Object.entries(mockDispatch.headers)) {
-    const header = headers instanceof HeadersList ? headers.get(matchHeaderName) : headers[matchHeaderName]
+    const header = headers[kHeadersList] ? headers.get(matchHeaderName) : headers[matchHeaderName]
```
Comment (ronag) [inline lib/mock/mock-utils.js:36]:
```suggestion
    const header = headers.get ? headers.get(matchHeaderName) : headers[matchHeaderName]
```

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**